File: scala-client/src/main/scala/com/couchbase/client/scala/manager/query/AsyncQueryIndexManager.scala
/*
 * Copyright (c) 2019 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.couchbase.client.scala.manager.query

import com.couchbase.client.core.Core
import com.couchbase.client.core.api.manager._
import com.couchbase.client.core.endpoint.http.CoreCommonOptions
import com.couchbase.client.core.error.InvalidArgumentException
import com.couchbase.client.core.manager.CoreQueryIndexManager
import com.couchbase.client.core.retry.{BestEffortRetryStrategy, RetryStrategy}
import com.couchbase.client.scala.AsyncCluster
import com.couchbase.client.scala.implicits.Codec
import com.couchbase.client.scala.util.DurationConversions._
import com.couchbase.client.scala.util.FutureConversions

import java.{lang, util}
import scala.compat.java8.OptionConverters._
import scala.concurrent.duration.Duration
import scala.concurrent.{ExecutionContext, Future}
import scala.jdk.CollectionConverters._

/** Allows query indexes to be managed.
  *
  * Operations take a bucketName, scopeName and collectionName.
  *
  * If only bucketName is provided, the indexes affected will be those on the bucket's default scope and collection.
  * If bucketName and scopeName are provided, the indexes affected will be all those on collections under that scope.
  * If bucketName, scopeName and collectionName are provided, the affected fetched will be on that specific collection.
  *
  * @define Timeout        when the operation will timeout.  This will default to `timeoutConfig().managementTimeout
  *                        ()` in the
  *                        provided [[com.couchbase.client.scala.env.ClusterEnvironment]].
  * @define RetryStrategy  provides some control over how the SDK handles failures.  Will default to `retryStrategy()`
  *                        in the provided [[com.couchbase.client.scala.env.ClusterEnvironment]].
  * @define ScopeName      if specified, this operation will work on a given [[com.couchbase.client.scala.Scope]] and
  *                        [[com.couchbase.client.scala.Collection]].  If specified, `collectionName` must also be specified.
  * @define CollectionName if specified, this operation will work on a given [[com.couchbase.client.scala.Scope]] and
  *                        [[com.couchbase.client.scala.Collection]].  If specified, `scopeName` must also be specified.
  */
class AsyncQueryIndexManager(private[scala] val cluster: AsyncCluster)(
    implicit val ec: ExecutionContext
) {
  private[scala] val internal = new CoreQueryIndexManager(
    cluster.couchbaseOps.queryOps(),
    cluster.core.coreResources.requestTracer
  )

  private[scala] val DefaultTimeout =
    cluster.couchbaseOps.environment.timeoutConfig.managementTimeout
  private[scala] val DefaultRetryStrategy = cluster.couchbaseOps.environment.retryStrategy

  /** Gets all indexes.
    *
    * If only bucketName is provided, all indexes for that bucket will be fetched, for all scopes and collections.
    * If bucketName and scopeName are provided, the indexes fetched will be all those on collections under that scope.
    * If bucketName, scopeName and collectionName are provided, the indexes fetched will be on that specific collection.
    *
    * @param bucketName     the bucket to get indexes on
    * @param timeout        $Timeout
    * @param retryStrategy  $RetryStrategy
    * @param scopeName      if specified, the indexes fetched will be limited to those on collections under this [[com.couchbase.client.scala.Scope]]
    * @param collectionName if specified, the indexes fetched will be limited to those on this specific [[com.couchbase.client.scala.Collection]]
    */
  def getAllIndexes(
      bucketName: String,
      timeout: Duration = DefaultTimeout,
      retryStrategy: RetryStrategy = DefaultRetryStrategy,
      @deprecated("Users should use `collection.queryIndexes()` instead") scopeName: Option[
        String
      ] = None,
      @deprecated("Users should use `collection.queryIndexes()` instead") collectionName: Option[
        String
      ] = None
  ): Future[collection.Seq[QueryIndex]] = {
    val sn = scopeName
    val cn = collectionName

    FutureConversions
      .javaCFToScalaFutureMappingExceptions(
        internal.getAllIndexes(
          bucketName,
          new CoreGetAllQueryIndexesOptions {
            override def scopeName(): String = sn.orNull

            override def collectionName(): String = cn.orNull

            override def commonOptions(): CoreCommonOptions =
              makeCommonOptions(timeout, retryStrategy)
          }
        )
      )
      .map(
        result =>
          result.asScala.map(
            i =>
              QueryIndex(
                i.name,
                Some(i.primary),
                Option(i.`type`),
                i.state,
                i.keyspace,
                i.indexKey.asScala.toSeq.map(v => v.textValue),
                i.condition.asScala,
                i.partition.asScala,
                Option(i.bucketName),
                i.scopeName.asScala
              )
          )
      )
  }

  /** Creates a new query index with the specified parameters.
    *
    * @param bucketName     the bucket to create the index on.
    * @param indexName      the name of the index.
    * @param ignoreIfExists if an index with the same name already exists, the operation will fail.
    * @param numReplicas    how many replicas of the index to create.
    * @param deferred       set to true to defer building the index until [[buildDeferredIndexes]] is called.  This can
    *                       provide improved performance when creating multiple indexes.
    * @param timeout        $Timeout
    * @param retryStrategy  $RetryStrategy
    * @param scopeName      $ScopeName
    * @param collectionName $CollectionName
    */
  def createIndex(
      bucketName: String,
      indexName: String,
      fields: Iterable[String],
      ignoreIfExists: Boolean = false,
      numReplicas: Option[Int] = None,
      deferred: Option[Boolean] = None,
      timeout: Duration = DefaultTimeout,
      retryStrategy: RetryStrategy = DefaultRetryStrategy,
      @deprecated("Users should use `collection.queryIndexes()` instead") scopeName: Option[
        String
      ] = None,
      @deprecated("Users should use `collection.queryIndexes()` instead") collectionName: Option[
        String
      ] = None
  ): Future[Unit] = {
    val _deferred       = deferred
    val _numReplicas    = numReplicas
    val _ignoreIfExists = ignoreIfExists

    FutureConversions
      .javaCFToScalaFutureMappingExceptions(
        internal.createIndex(
          bucketName,
          indexName,
          fields.toSeq.asJava,
          new CoreCreateQueryIndexOptions {
            override def ignoreIfExists(): Boolean = _ignoreIfExists

            override def `with`(): util.Map[String, AnyRef] = null

            override def scopeAndCollection(): CoreScopeAndCollection =
              makeScopeAndCollection(scopeName, collectionName)

            override def commonOptions(): CoreCommonOptions =
              makeCommonOptions(timeout, retryStrategy)

            override def numReplicas(): Integer = _numReplicas.map(Integer.valueOf).orNull

            override def deferred(): lang.Boolean =
              _deferred.map(v => lang.Boolean.valueOf(v)).orNull
          }
        )
      )
      .map(_ => ())
  }

  /** Creates a new primary query index with the specified parameters.
    *
    * @param bucketName     the bucket to create the index on.
    * @param indexName      the name of the index.  If not set the server assigns the default primary index name.
    * @param ignoreIfExists if a primary index already exists, the operation will fail.
    * @param numReplicas    how many replicas of the index to create.
    * @param deferred       set to true to defer building the index until [[buildDeferredIndexes]] is called.  This can
    *                       provide improved performance when creating multiple indexes.
    * @param timeout        $Timeout
    * @param retryStrategy  $RetryStrategy
    * @param scopeName      $ScopeName
    * @param collectionName $CollectionName
    */
  def createPrimaryIndex(
      bucketName: String,
      indexName: Option[String] = None,
      ignoreIfExists: Boolean = false,
      numReplicas: Option[Int] = None,
      deferred: Option[Boolean] = None,
      timeout: Duration = DefaultTimeout,
      retryStrategy: RetryStrategy = DefaultRetryStrategy,
      @deprecated("Users should use `collection.queryIndexes()` instead") scopeName: Option[
        String
      ] = None,
      @deprecated("Users should use `collection.queryIndexes()` instead") collectionName: Option[
        String
      ] = None
  ): Future[Unit] = {
    val _deferred       = deferred
    val _numReplicas    = numReplicas
    val _ignoreIfExists = ignoreIfExists
    val _indexName      = indexName

    FutureConversions
      .javaCFToScalaFutureMappingExceptions(
        internal.createPrimaryIndex(
          bucketName,
          new CoreCreatePrimaryQueryIndexOptions {
            override def ignoreIfExists(): Boolean = _ignoreIfExists

            override def `with`(): util.Map[String, AnyRef] = null

            override def scopeAndCollection(): CoreScopeAndCollection =
              makeScopeAndCollection(scopeName, collectionName)

            override def commonOptions(): CoreCommonOptions =
              makeCommonOptions(timeout, retryStrategy)

            override def numReplicas(): Integer = _numReplicas.map(Integer.valueOf).orNull

            override def deferred(): lang.Boolean =
              _deferred.map(v => lang.Boolean.valueOf(v)).orNull

            override def indexName(): String = _indexName.orNull
          }
        )
      )
      .map(_ => ())
  }

  /** Drops an existing index.
    *
    * @param bucketName        the bucket to remove the index from.
    * @param indexName         the name of the index.
    * @param ignoreIfNotExists sets whether the operation should fail if the index does not exists
    * @param timeout           $Timeout
    * @param retryStrategy     $RetryStrategy
    * @param scopeName         $ScopeName
    * @param collectionName    $CollectionName
    */
  def dropIndex(
      bucketName: String,
      indexName: String,
      ignoreIfNotExists: Boolean = false,
      timeout: Duration = DefaultTimeout,
      retryStrategy: RetryStrategy = DefaultRetryStrategy,
      @deprecated("Users should use `collection.queryIndexes()` instead") scopeName: Option[
        String
      ] = None,
      @deprecated("Users should use `collection.queryIndexes()` instead") collectionName: Option[
        String
      ] = None
  ): Future[Unit] = {
    val _ignoreIfNotExists = ignoreIfNotExists

    FutureConversions
      .javaCFToScalaFutureMappingExceptions(
        internal.dropIndex(
          bucketName,
          indexName,
          new CoreDropQueryIndexOptions {
            override def ignoreIfNotExists(): Boolean = _ignoreIfNotExists

            override def scopeAndCollection(): CoreScopeAndCollection =
              makeScopeAndCollection(scopeName, collectionName)

            override def commonOptions(): CoreCommonOptions =
              makeCommonOptions(timeout, retryStrategy)
          }
        )
      )
      .map(_ => ())
  }

  /** Drops an existing primary index.
    *
    * @param bucketName        the bucket to remove the index from.
    * @param ignoreIfNotExists sets whether the operation should fail if the index does not exists
    * @param timeout           $Timeout
    * @param retryStrategy     $RetryStrategy
    * @param scopeName         $ScopeName
    * @param collectionName    $CollectionName
    */
  def dropPrimaryIndex(
      bucketName: String,
      ignoreIfNotExists: Boolean = false,
      timeout: Duration = DefaultTimeout,
      retryStrategy: RetryStrategy = DefaultRetryStrategy,
      @deprecated("Users should use `collection.queryIndexes()` instead") scopeName: Option[
        String
      ] = None,
      @deprecated("Users should use `collection.queryIndexes()` instead") collectionName: Option[
        String
      ] = None
  ): Future[Unit] = {
    val _ignoreIfNotExists = ignoreIfNotExists

    FutureConversions
      .javaCFToScalaFutureMappingExceptions(
        internal.dropPrimaryIndex(
          bucketName,
          new CoreDropPrimaryQueryIndexOptions {
            override def ignoreIfNotExists(): Boolean = _ignoreIfNotExists

            override def scopeAndCollection(): CoreScopeAndCollection =
              makeScopeAndCollection(scopeName, collectionName)

            override def commonOptions(): CoreCommonOptions =
              makeCommonOptions(timeout, retryStrategy)
          }
        )
      )
      .map(_ => ())
  }

  /** Polls the specified indexes until they are all online.
    *
    * @param bucketName        the bucket to remove the index from.
    * @param indexNames        the indexes to poll.
    * @param watchPrimary      include the bucket's primary index.  If the bucket has no primary index, the operation
    *                          will fail with `IndexNotFoundException`
    * @param timeout           when the operation will timeout.
    * @param retryStrategy     $RetryStrategy
    * @param scopeName         $ScopeName
    * @param collectionName    $CollectionName
    */
  def watchIndexes(
      bucketName: String,
      indexNames: Iterable[String],
      timeout: Duration,
      watchPrimary: Boolean = false,
      retryStrategy: RetryStrategy = DefaultRetryStrategy,
      @deprecated("Users should use `collection.queryIndexes()` instead") scopeName: Option[
        String
      ] = None,
      @deprecated("Users should use `collection.queryIndexes()` instead") collectionName: Option[
        String
      ] = None
  ): Future[Unit] = {
    val _watchPrimary = watchPrimary

    FutureConversions
      .javaCFToScalaFutureMappingExceptions(
        internal.watchIndexes(
          bucketName,
          indexNames.toSeq.asJava,
          timeout,
          new CoreWatchQueryIndexesOptions {
            override def watchPrimary(): Boolean = _watchPrimary

            override def scopeAndCollection(): CoreScopeAndCollection =
              makeScopeAndCollection(scopeName, collectionName)

            override def commonOptions(): CoreCommonOptions =
              makeCommonOptions(timeout, retryStrategy)
          }
        )
      )
      .map(_ => ())
  }

  /** Build all deferred indexes.
    *
    * @param bucketName        the bucket to build indexes on.
    * @param timeout           $Timeout
    * @param retryStrategy     $RetryStrategy
    * @param scopeName         $ScopeName
    * @param collectionName    $CollectionName
    */
  def buildDeferredIndexes(
      bucketName: String,
      timeout: Duration = DefaultTimeout,
      retryStrategy: RetryStrategy = DefaultRetryStrategy,
      @deprecated("Users should use `collection.queryIndexes()` instead") scopeName: Option[
        String
      ] = None,
      @deprecated("Users should use `collection.queryIndexes()` instead") collectionName: Option[
        String
      ] = None
  ): Future[Unit] = {
    FutureConversions
      .javaCFToScalaFutureMappingExceptions(
        internal.buildDeferredIndexes(
          bucketName,
          new CoreBuildQueryIndexOptions {
            override def scopeAndCollection(): CoreScopeAndCollection =
              makeScopeAndCollection(scopeName, collectionName)

            override def commonOptions(): CoreCommonOptions =
              makeCommonOptions(timeout, retryStrategy)
          }
        )
      )
      .map(_ => ())
  }

  def makeScopeAndCollection(
      scopeName: Option[String] = None,
      collectionName: Option[String] = None
  ): CoreScopeAndCollection = {
    if (scopeName.isDefined && collectionName.isDefined) {
      new CoreScopeAndCollection(scopeName.get, collectionName.get)
    } else if (scopeName.isDefined || collectionName.isDefined) {
      // Will only be thrown inside core-io, so will be safely wrapped into a Try/Future/SMono during conversion
      throw new InvalidArgumentException(
        "If either scopeName or collectionName is provided, both must be.  However, it's recommended that users use CollectionQueryIndexManager instead",
        null,
        null
      )
    } else {
      null
    }
  }

  def makeCommonOptions(
      timeout: Duration = DefaultTimeout,
      retryStrategy: RetryStrategy = DefaultRetryStrategy
  ): CoreCommonOptions = {
    CoreCommonOptions.of(
      if (timeout == DefaultTimeout) null else timeout,
      if (retryStrategy == DefaultRetryStrategy) null else retryStrategy,
      null
    )
  }
}

case class QueryIndex(
    name: String,
    private val is_primary: Option[Boolean],
    private val using: Option[String],
    state: String,
    private val keyspace_id: String,
    private val index_key: Seq[String],
    condition: Option[String],
    partition: Option[String],
    private val bucket_id: Option[String],
    private val scope_id: Option[String]
) {

  /** If this is a collection-level index then this will be the collection's name, otherwise the bucket name. */
  def keyspaceId: String = keyspace_id

  /** If this is a collection-level index then this will be the collection's scope's name, otherwise empty. */
  def scopeName: Option[String] = scope_id

  /** The name of the bucket the index is on. */
  def bucketName: String = scope_id match {
    case Some(_) => bucket_id.get // .get is ok, must be present
    case _       => keyspace_id
  }

  /** If this is a collection-level index then this will be the collection's name, otherwise empty. */
  def collectionName: Option[String] = scope_id match {
    case Some(_) => Some(keyspace_id)
    case _       => None
  }

  def isPrimary: Boolean = is_primary.getOrElse(false)

  def indexKey: Seq[String] = index_key

  def typ: QueryIndexType = using match {
    case Some("gsi") => QueryIndexType.GSI
    case _           => QueryIndexType.View
  }
}

object QueryIndex {
  implicit val codec: Codec[QueryIndex] = Codec.codec[QueryIndex]
}

sealed trait QueryIndexType

object QueryIndexType {
  case object View extends QueryIndexType

  case object GSI extends QueryIndexType
}

File: java-client/src/main/java/com/couchbase/client/java/manager/query/AsyncQueryIndexManager.java
/*
 * Copyright 2019 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.couchbase.client.java.manager.query;

import com.couchbase.client.core.annotation.Stability;
import com.couchbase.client.core.error.CouchbaseException;
import com.couchbase.client.core.error.IndexExistsException;
import com.couchbase.client.core.error.IndexFailureException;
import com.couchbase.client.core.error.IndexNotFoundException;
import com.couchbase.client.core.manager.CoreQueryIndexManager;
import com.couchbase.client.java.AsyncCluster;

import java.time.Duration;
import java.util.Collection;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

import static com.couchbase.client.java.manager.query.BuildQueryIndexOptions.buildDeferredQueryIndexesOptions;
import static com.couchbase.client.java.manager.query.CreatePrimaryQueryIndexOptions.createPrimaryQueryIndexOptions;
import static com.couchbase.client.java.manager.query.CreateQueryIndexOptions.createQueryIndexOptions;
import static com.couchbase.client.java.manager.query.DropPrimaryQueryIndexOptions.dropPrimaryQueryIndexOptions;
import static com.couchbase.client.java.manager.query.DropQueryIndexOptions.dropQueryIndexOptions;
import static com.couchbase.client.java.manager.query.GetAllQueryIndexesOptions.getAllQueryIndexesOptions;
import static com.couchbase.client.java.manager.query.WatchQueryIndexesOptions.watchQueryIndexesOptions;
import static java.util.Objects.requireNonNull;

/**
 * Performs management operations on query indexes.
 */
public class AsyncQueryIndexManager {

  private final CoreQueryIndexManager internal;

  /**
   * Creates a new {@link AsyncQueryIndexManager}.
   * <p>
   * This API is not intended to be called by the user directly, use {@link AsyncCluster#queryIndexes()}
   * instead.
   *
   * @param cluster the async cluster to perform the queries on.
   */
  @Stability.Internal
  public AsyncQueryIndexManager(final AsyncCluster cluster) {
    this.internal = new CoreQueryIndexManager(requireNonNull(cluster).core());
  }

  /**
   * Creates a named query index.
   * <p>
   * By default, this method will create an index on the bucket. If an index needs to be created on a collection,
   * both {@link CreateQueryIndexOptions#scopeName(String)} and {@link CreateQueryIndexOptions#collectionName(String)}
   * must be set.
   *
   * @param bucketName the name of the bucket to create the index on.
   * @param indexName the name of the query index.
   * @param fields the collection of fields that are part of the index.
   * @return a {@link CompletableFuture} completing when the operation is applied or failed with an error.
   * @throws IndexFailureException (async) if creating the index failed (see reason for details).
   * @throws IndexExistsException (async) if an index already exists with the given name on the keyspace.
   * @throws CouchbaseException (async) if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> createIndex(final String bucketName, final String indexName,
                                             final Collection<String> fields) {
    return createIndex(bucketName, indexName, fields, createQueryIndexOptions());
  }

  /**
   * Creates a named query index with custom options.
   * <p>
   * By default, this method will create an index on the bucket. If an index needs to be created on a collection,
   * both {@link CreateQueryIndexOptions#scopeName(String)} and {@link CreateQueryIndexOptions#collectionName(String)}
   * must be set.
   *
   * @param bucketName the name of the bucket to create the index on.
   * @param indexName the name of the query index.
   * @param fields the collection of fields that are part of the index.
   * @param options the custom options to apply.
   * @return a {@link CompletableFuture} completing when the operation is applied or failed with an error.
   * @throws IndexFailureException (async) if creating the index failed (see reason for details).
   * @throws IndexExistsException (async) if an index already exists with the given name on the keyspace.
   * @throws CouchbaseException (async) if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> createIndex(final String bucketName, final String indexName,
                                             final Collection<String> fields, final CreateQueryIndexOptions options) {
    return internal.createIndex(bucketName, indexName, fields, options.build());
  }

  /**
   * Creates a primary query index.
   * <p>
   * By default, this method will create an index on the bucket. If an index needs to be created on a collection,
   * both {@link CreatePrimaryQueryIndexOptions#scopeName(String)} and
   * {@link CreatePrimaryQueryIndexOptions#collectionName(String)} must be set.
   *
   * @param bucketName the name of the bucket to create the index on.
   * @return a {@link CompletableFuture} completing when the operation is applied or failed with an error.
   * @throws IndexFailureException (async) if creating the index failed (see reason for details).
   * @throws IndexExistsException (async) if an index already exists with the given name on the keyspace.
   * @throws CouchbaseException (async) if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> createPrimaryIndex(final String bucketName) {
    return createPrimaryIndex(bucketName, createPrimaryQueryIndexOptions());
  }

  /**
   * Creates a primary query index with custom options.
   * <p>
   * By default, this method will create an index on the bucket. If an index needs to be created on a collection,
   * both {@link CreatePrimaryQueryIndexOptions#scopeName(String)} and
   * {@link CreatePrimaryQueryIndexOptions#collectionName(String)} must be set.
   *
   * @param bucketName the name of the bucket to create the index on.
   * @param options the custom options to apply.
   * @return a {@link CompletableFuture} completing when the operation is applied or failed with an error.
   * @throws IndexFailureException (async) if creating the index failed (see reason for details).
   * @throws IndexExistsException (async) if an index already exists with the given name on the keyspace.
   * @throws CouchbaseException (async) if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> createPrimaryIndex(final String bucketName,
                                                    final CreatePrimaryQueryIndexOptions options) {
    return internal.createPrimaryIndex(bucketName, options.build());
  }

  /**
   * Fetches all indexes from the bucket.
   * <p>
   * By default, this method will fetch all index on the bucket. If the indexes should be loaded for a collection,
   * both {@link GetAllQueryIndexesOptions#scopeName(String)} and
   * {@link GetAllQueryIndexesOptions#collectionName(String)} must be set. If all indexes for a scope should be loaded,
   * only the {@link GetAllQueryIndexesOptions#scopeName(String)} can be set.
   *
   * @param bucketName the name of the bucket to load the indexes from.
   * @return a {@link CompletableFuture} completing with a list of (potentially empty) indexes or failed with an error.
   * @throws CouchbaseException (async) if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<List<QueryIndex>> getAllIndexes(final String bucketName) {
    return getAllIndexes(bucketName, getAllQueryIndexesOptions());
  }

  /**
   * Fetches all indexes from the bucket with custom options.
   * <p>
   * By default, this method will fetch all index on the bucket. If the indexes should be loaded for a collection,
   * both {@link GetAllQueryIndexesOptions#scopeName(String)} and
   * {@link GetAllQueryIndexesOptions#collectionName(String)} must be set. If all indexes for a scope should be loaded,
   * only the {@link GetAllQueryIndexesOptions#scopeName(String)} can be set.
   *
   * @param bucketName the name of the bucket to load the indexes from.
   * @param options the custom options to apply.
   * @return a {@link CompletableFuture} completing with a list of (potentially empty) indexes or failed with an error.
   * @throws CouchbaseException (async) if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<List<QueryIndex>> getAllIndexes(final String bucketName,
                                                           final GetAllQueryIndexesOptions options) {
    return internal.getAllIndexes(bucketName, options.build())
            .thenApply(rows -> rows.stream()
                    .map(row -> new QueryIndex(row))
                    .collect(Collectors.toList()));
  }

  /**
   * Drops the primary index from a bucket.
   * <p>
   * By default, this method will drop the primary index on the bucket. If the index should be dropped on a collection,
   * both {@link DropPrimaryQueryIndexOptions#scopeName(String)} and
   * {@link DropPrimaryQueryIndexOptions#collectionName(String)} must be set.
   *
   * @param bucketName the name of the bucket to drop the indexes from.
   * @return a {@link CompletableFuture} completing when the operation is applied or failed with an error.
   * @throws IndexNotFoundException (async) if the index does not exist.
   * @throws IndexFailureException (async) if dropping the index failed (see reason for details).
   * @throws CouchbaseException (async) if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> dropPrimaryIndex(final String bucketName) {
    return dropPrimaryIndex(bucketName, dropPrimaryQueryIndexOptions());
  }

  /**
   * Drops the primary index from a bucket with custom options.
   * <p>
   * By default, this method will drop the primary index on the bucket. If the index should be dropped on a collection,
   * both {@link DropPrimaryQueryIndexOptions#scopeName(String)} and
   * {@link DropPrimaryQueryIndexOptions#collectionName(String)} must be set.
   *
   * @param bucketName the name of the bucket to drop the indexes from.
   * @param options the custom options to apply.
   * @return a {@link CompletableFuture} completing when the operation is applied or failed with an error.
   * @throws IndexNotFoundException (async) if the index does not exist.
   * @throws IndexFailureException (async) if dropping the index failed (see reason for details).
   * @throws CouchbaseException (async) if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> dropPrimaryIndex(final String bucketName, final DropPrimaryQueryIndexOptions options) {
    return internal.dropPrimaryIndex(bucketName, options.build());
  }

  /**
   * Drops a query index from a bucket.
   * <p>
   * By default, this method will drop the index on the bucket. If the index should be dropped on a collection,
   * both {@link DropQueryIndexOptions#scopeName(String)} and
   * {@link DropQueryIndexOptions#collectionName(String)} must be set.
   *
   * @param bucketName the name of the bucket to drop the indexes from.
   * @param indexName the name of the index top drop.
   * @return a {@link CompletableFuture} completing when the operation is applied or failed with an error.
   * @throws IndexNotFoundException (async) if the index does not exist.
   * @throws IndexFailureException (async) if dropping the index failed (see reason for details).
   * @throws CouchbaseException (async) if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> dropIndex(final String bucketName, final String indexName) {
    return dropIndex(bucketName, indexName, dropQueryIndexOptions());
  }

  /**
   * Drops a query index from a bucket with custom options.
   * <p>
   * By default, this method will drop the index on the bucket. If the index should be dropped on a collection,
   * both {@link DropQueryIndexOptions#scopeName(String)} and
   * {@link DropQueryIndexOptions#collectionName(String)} must be set.
   *
   * @param bucketName the name of the bucket to drop the indexes from.
   * @param indexName the name of the index top drop.
   * @param options the custom options to apply.
   * @return a {@link CompletableFuture} completing when the operation is applied or failed with an error.
   * @throws IndexNotFoundException (async) if the index does not exist.
   * @throws IndexFailureException (async) if dropping the index failed (see reason for details).
   * @throws CouchbaseException (async) if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> dropIndex(final String bucketName, final String indexName,
                                           final DropQueryIndexOptions options) {
    return internal.dropIndex(bucketName, indexName, options.build());
  }

  /**
   * Builds all currently deferred indexes in the bucket's default collection.
   * <p>
   * To target a different collection, see {@link #buildDeferredIndexes(String, BuildQueryIndexOptions)}.
   *
   * @param bucketName the name of the bucket to build deferred indexes for.
   * @return a {@link CompletableFuture} completing when the operation is applied or failed with an error.
   * @throws CouchbaseException (async) if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> buildDeferredIndexes(final String bucketName) {
    return buildDeferredIndexes(bucketName, buildDeferredQueryIndexesOptions());
  }

  /**
   * Builds all currently deferred indexes in a collection.
   * <p>
   * By default, this method targets the bucket's default collection.
   * To target a different collection, specify both 
   * {@link BuildQueryIndexOptions#scopeName(String)} and
   * {@link BuildQueryIndexOptions#collectionName(String)}.
   *
   * @param bucketName the name of the bucket to build deferred indexes for.
   * @param options the custom options to apply.
   * @return a {@link CompletableFuture} completing when the operation is applied or failed with an error.
   * @throws CouchbaseException (async) if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> buildDeferredIndexes(final String bucketName, final BuildQueryIndexOptions options) {
    return internal.buildDeferredIndexes(bucketName, options.build());
  }

  /**
   * Watches/Polls indexes until they are online.
   * <p>
   * By default, this method will watch the indexes on the bucket. If the indexes should be watched on a collection,
   * both {@link WatchQueryIndexesOptions#scopeName(String)} and
   * {@link WatchQueryIndexesOptions#collectionName(String)} must be set.
   *
   * @param bucketName the name of the bucket where the indexes should be watched.
   * @param indexNames the names of the indexes to watch.
   * @param timeout the maximum amount of time the indexes should be watched.
   * @return a {@link CompletableFuture} completing when the operation is applied or failed with an error.
   * @throws CouchbaseException (async) if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> watchIndexes(final String bucketName, final Collection<String> indexNames,
                                              final Duration timeout) {
    return watchIndexes(bucketName, indexNames, timeout, watchQueryIndexesOptions());
  }

  /**
   * Watches/Polls indexes until they are online with custom options.
   * <p>
   * By default, this method will watch the indexes on the bucket. If the indexes should be watched on a collection,
   * both {@link WatchQueryIndexesOptions#scopeName(String)} and
   * {@link WatchQueryIndexesOptions#collectionName(String)} must be set.
   *
   * @param bucketName the name of the bucket where the indexes should be watched.
   * @param indexNames the names of the indexes to watch.
   * @param timeout the maximum amount of time the indexes should be watched.
   * @param options the custom options to apply.
   * @return a {@link CompletableFuture} completing when the operation is applied or failed with an error.
   * @throws CouchbaseException (async) if any other generic unhandled/unexpected errors.
   */
  public CompletableFuture<Void> watchIndexes(final String bucketName, final Collection<String> indexNames,
                                              final Duration timeout, final WatchQueryIndexesOptions options) {
    return internal.watchIndexes(bucketName, indexNames, timeout, options.build());
  }
}
